pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -83,93 +87,132 @@ func (c *ClusterMemberController) sync(ctx context.Context, syncCtx factory.Sync
 
-func (c *ClusterMemberController) getEtcdPodToAddToMembership(ctx context.Context) (*corev1.Pod, error) {
-	// list etcd member pods
-	pods, err := c.podLister.List(labels.Set{"app": "etcd"}.AsSelector())
-	if err != nil {
-		return nil, err
-	}
+// getEtcdPeerURLToScale returns a PeerURL of the next etcd instance to be scaled up in the cluster. If
+// no member is found to scale up return empty string.
 
-	// go through the list of all pods, pick one peerFQDN to return from unready pods
-	// and collect dns resolution errors on the way.
-	for _, pod := range pods {
-		if !strings.HasPrefix(pod.Name, "etcd-") {
-			continue
-		}
-		isEtcdContainerRunning, isEtcdContainerReady := false, false
-		for _, containerStatus := range pod.Status.ContainerStatuses {
-			if containerStatus.Name != "etcd" {
-				continue
-			}
-			// set running and ready flags
-			isEtcdContainerRunning = containerStatus.State.Running != nil
-			isEtcdContainerReady = containerStatus.Ready
-			break
-		}
-		if !isEtcdContainerRunning || isEtcdContainerReady {
-			continue
-		}
+// Change: The understanding of
 
-		// now check to see if this member is already part of the quorum.  This logically requires being able to map every
-		// type of member name we have ever created.  The most important for now is the nodeName.
-		etcdMember, err := c.etcdClient.GetMember(ctx, pod.Spec.NodeName)
-		switch {
-		case apierrors.IsNotFound(err):
-			return pod, nil
-		case err != nil:
-			return nil, err
-		default:
-			klog.Infof("skipping unready pod %q because it is already an etcd member: %#v", pod.Name, etcdMember)
+func (c *ClusterMemberController) getEtcdPeerURLToScale(ctx context.Context) (string, error) {
+	nodes, err := c.nodeLister.List(labels.SelectorFromSet(labels.Set{masterLabel: ""}))
+	if err != nil {
+		return "", err
+	}
+	if c.replicaCount == 0 {
+		replicaCount, err := ceohelpers.GetMastersReplicaCount(ctx, c.kubeClient)
+		if err != nil {
+			return "", fmt.Errorf("failed to get control-plane replica count: %w", err)
 		}
+		c.replicaCount = replicaCount
 	}
-	return nil, nil
-}
+	desiredMemberCount := c.replicaCount
 
-// getValidPodFQDNToScale goes through the list on unready pods and
-// returns a resolvable  podFQDN. If none of the DNSes are available
-// yet it will return collected errors.
-func (c *ClusterMemberController) getEtcdPeerHostToScale(podToAdd *corev1.Pod) (string, error) {
-	network, err := c.networkLister.Get("cluster")
+	// Sort nodes by created timestamp.
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].CreationTimestamp.Before(&nodes[j].CreationTimestamp)
+	})
+
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
 		return "", err
 	}
-	node, err := c.nodeLister.Get(podToAdd.Spec.NodeName)
+
+	currentMemberCount := len(nodes)
+	maxMemberCount := desiredMemberCount + 1 // Surge up
+	if currentMemberCount == maxMemberCount {
+		klog.Warningf("Skipping scale up etcd membership at maximum size: %d", desiredMemberCount+1)
```

Comment:
@hexfusion During a manual test run I see that when I add a new node, `currentMemberCount=4` and `desiredMemberCount=3` since that's sourced from the `cluster-config-v1` configmap's `data.install-config.controlPlane.replicas` which is 3.

So the clustermember controller here will keep spinning on `currentMemberCount == maxMemberCount` even though we're not really at the max membership size yet.

Also at this point seems like there is no etcd static pod on the new node as well:
```yaml
kind: Etcd
metadata:
  name: cluster
...
status:
  conditions:
  - lastTransitionTime: "2021-11-09T20:26:46Z"
    message: 3 nodes are active; 1 nodes are at revision 0; 3 nodes are at revision
      6
    status: "True"
    type: StaticPodsAvailable
  - lastTransitionTime: "2021-11-10T05:22:11Z"
    message: 1 nodes are at revision 0; 3 nodes are at revision 6
    status: "True"
    type: NodeInstallerProgressing
    ...
    nodeStatuses:
  - currentRevision: 6
    nodeName: ip-10-0-136-64.us-west-1.compute.internal
  - currentRevision: 6
    nodeName: ip-10-0-227-172.us-west-1.compute.internal
  - currentRevision: 6
    nodeName: ip-10-0-186-101.us-west-1.compute.internal
  - currentRevision: 0
    nodeName: ip-10-0-143-124.us-west-1.compute.internal
```

It's only when I update the configmap `controlPlane.replicas=4` (and kick the etcd-operator to restart since it's not noticing the change for some reason), does it actually progress past this point, and the cluster settles to 4 available members with all the nodes at revision 7.

But I don't think we want admins to update that configmap since the desired controlplane size should be 3 at the end.


So (other than understanding why the static pod doesn't initially show up on the new node), I think `currentMemberCount` here should be the current number of voting members and not the number of nodes.
E.g we could add 3 new nodes at which point our  `currentMemberCount=6 > maxMemberCount=4`
```suggestion
	currentMemberCount := len(members)
	maxMemberCount := desiredMemberCount + 1 // Surge up
	if currentMemberCount == maxMemberCount {
		klog.Warningf("Skipping scale up etcd membership at maximum size: %d", desiredMemberCount+1)
```

Reply:
Although that will only allow us to scale up and down serially, 1 at a time, `3->4->3->4->3->4->3` rather than all at once.

Reply:
That is the goal post bootstrap while we can add new learners in parallel promotion in this circumstance would be serial.  Yes and in general we are moving away from nodes to members so the change makes sense.